File: src/services/lessonProgressService.ts
import apiService from "@/apis/apiService";
import {
    LessonProgressRequest,
    LessonProgressResponse,
    CourseProgressSummaryResponse
} from "@/model/progress/LessonProgress";

const PROGRESS_BASE_ENDPOINT = "/progress";

const lessonProgressService = {
    /**
     * Update or create lesson progress
     */
    updateProgress: (
        request: LessonProgressRequest
    ): Promise<LessonProgressResponse> => {
        return apiService.post<LessonProgressResponse>(
            PROGRESS_BASE_ENDPOINT,
            request
        );
    },

    /**
     * Get all progress for a course
     */
    getCourseProgress: (
        courseId: string
    ): Promise<LessonProgressResponse[]> => {
        return apiService.get<LessonProgressResponse[]>(
            `${PROGRESS_BASE_ENDPOINT}/course/${courseId}`
        );
    },

    /**
     * Get course progress summary with stats and last accessed lesson
     */
    getCourseSummary: (
        courseId: string
    ): Promise<CourseProgressSummaryResponse> => {
        return apiService.get<CourseProgressSummaryResponse>(
            `${PROGRESS_BASE_ENDPOINT}/course/${courseId}/summary`
        );
    },

    /**
     * Get progress for a specific lesson
     */
    getLessonProgress: (
        courseId: string,
        lessonId: string
    ): Promise<LessonProgressResponse | null> => {
        return apiService.get<LessonProgressResponse>(
            `${PROGRESS_BASE_ENDPOINT}/course/${courseId}/lesson/${lessonId}`
        ).catch(() => null); // Return null if no progress exists
    }
};

export default lessonProgressService;

File: src/main/java/com/truongsonkmhd/unetistudy/service/LessonProgressService.java
package com.truongsonkmhd.unetistudy.service;

import com.truongsonkmhd.unetistudy.dto.progress_dto.CourseProgressSummaryResponse;
import com.truongsonkmhd.unetistudy.dto.progress_dto.LessonProgressRequest;
import com.truongsonkmhd.unetistudy.dto.progress_dto.LessonProgressResponse;

import java.util.List;
import java.util.UUID;

/**
 * Service interface for managing lesson progress
 */
public interface LessonProgressService {

    /**
     * Update or create lesson progress for a user
     */
    LessonProgressResponse updateProgress(UUID userId, LessonProgressRequest request);

    /**
     * Get all progress for a user in a course
     */
    List<LessonProgressResponse> getCourseProgress(UUID userId, UUID courseId);

    /**
     * Get course progress summary with completion stats
     */
    CourseProgressSummaryResponse getCourseSummary(UUID userId, UUID courseId);

    /**
     * Get progress for a specific lesson
     */
    LessonProgressResponse getLessonProgress(UUID userId, UUID courseId, UUID lessonId);
}
